src/Features/CSharp/Portable/Completion/KeywordRecommenders/StackAllocKeywordRecommender.cs

```diff
@@ -16,20 +16,48 @@ public StackAllocKeywordRecommender()
 
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
-            // type t = |
-            var token = context.TargetToken;
-            if (token.IsUnsafeContext())
+            var node = context.TargetToken.Parent;
+
+            // In case of an empty string
+            if (node == null)
+            {
+                return false;
+            }
+
+            // After a cast
+            if (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) &&
+                (node.IsKind(SyntaxKind.ParenthesizedExpression) || node.IsKind(SyntaxKind.CastExpression)))
+            {
+                node = node.Parent;
+            }
+
+            // Inside a conditional expression: value ? stackalloc : stackalloc
+            while (node.IsKind(SyntaxKind.ConditionalExpression) &&
+                (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) || context.TargetToken.IsKind(SyntaxKind.QuestionToken) || context.TargetToken.IsKind(SyntaxKind.ColonToken)))
```

Comment:
i don't understand the close paren case.

Reply:
For cases like ` value ? (Span<int>stackalloc`
For things like this, we don't actually create the `WhenTrue` and `WhenFalse` nodes of the conditional yet.

---
In reply to: [156192660](https://github.com/dotnet/roslyn/pull/23655#discussion_r156192660) [](ancestors = 156192660)

Reply:
Do you mean ```value ? (Span<int>)stackalloc``` ?

If so, it's not clear to me why this isn't handled with this code:

```
if (context.TargetToken.IsKind(SyntaxKind.CloseParenToken) &&
 +                (node.IsKind(SyntaxKind.ParenthesizedExpression) || node.IsKind(SyntaxKind.CastExpression)))
```

Reply:
So, the best way to write this would be: 

1. if the target token is the ? or : of a ternay then the node should be the ternary
2. As long as we're the expression of a cast, or the true/false side of a ternary, walk upwards.
3. finally, see if we're on the right of an assignment or variable decl.